### DISCUSSION TITLE
What URL to code for an image in "public"

### DISCUSSION THREAD

**Discussion by belfasthenry** [mojolicious/mojo#1634] [Q&A] [answered]
My source tree looks (in part) like this:
<pre>
  MyApp
  +--bin
    +--myapp.pl
  +--public
     +--photos
       +--Sam
          +--somephoto.jpg </pre>

I can't find out how to code the `<img>` tag to contain a valid URL to that image.  All the questions I've found on the web say that files under "public" will be served out via the web pages (that is, "public" is the "docroot" in Apache terms); so I've coded in my template `%= image "photos/Sam/somephoto.jpg"` but the image doesn't load, and inspecting it in the browser shows "cannot load the image", despite the actual HTML source of the page containing a valid link `<img alt="somephoto.jpg" src="photos/Sam/somephoto.jpg">`

Pasting `http://127.0.0.1:3000/photos/Sam/somephoto.jpg` gives a "Page not found ... yet" message.  I've also tried `/photos ...` without success.

What am I doing wrong?

**Comment by jberger** [comment, marked as answer]
What do you get when you run `perl bin/myapp.pl eval -V 'app->static->paths'`?

**Reply by belfasthenry** [reply]
Aha ... I get
<pre>
[
  "/home/henry/myapp/bin/public"
]
</pre>
... which is unexpected.

**Reply by belfasthenry** [reply]
Hah!  That got me on the right track.  The "public" directory needs to be a **subdirectory of the directory that the executable is in** (`/home/henry/myapp/bin` in my case).  I think I'll eliminate bin and simply put the executable in the root directory of myapp.

I'm reluctant to blame the docs, but that's not what they say.  Or am I still missing something?

Anyway, the forums reign; thank you everbody.  (er, I should mark this "solved" but I can't see how)

**Reply by Grinnz** [reply]
Since your structure does not have any module files, I assume this is still a Lite app, in which case it indeed does not expect to be in a subdirectory.

**Reply by belfasthenry** [reply]
Yes, that's true, although I simplified the structure for the sake of this post: I do have a "lib" directory with a module in it which implements a lot of my functions.

Having understood what I've done, and maybe got some glimmer about what I ought to be doing, can someone recommend where I might read authoritatively about the differences between Mojolicious and Mojolicious::Lite?  I've been bitten by this one, and I guess there are others.

**Reply by Grinnz** [reply]
The Growing guide details the changes necessary to turn a Lite app into a full app. But it's really simpler than you may think; the Lite app is simply a tiny wrapper of a full app that provides a DSL and allows you to use the script file as the application file, and as a result expects that file to be in the project root.

**Reply by Grinnz** [reply]
By the way, you can click the check-mark in the upper right of this answer to mark it as the answer.

**Reply by belfasthenry** [reply]
For example, the Tutorial document says it's an "example-driven introduction to the wonders of Mojolicious::Lite.".  Then under "Static files" the same document says "External static files ... will be served automatically from a public directory if it exists."  But, as I've just discovered the hard way, for a ::Lite application that's not the case.

**Reply by Grinnz** [reply]
All of that is accurate. Your problem was a misplaced Lite script file.

**Comment by Grinnz** [comment]
Short answer: `/photos/Sam/somephoto.jpg`

Long answer: There are three kinds of URLs, absolute, host-relative, and page-relative. You have provided a page-relative URL, so the browser will construct the URL to request depending on the current page path it's on. You want a host-relative URL which starts with a slash, so the browser will request it relative to the root of your application.

**Reply by belfasthenry** [reply]
I tried that.  I probably didn't phrase my original question sufficiently clearly, but I did say towards the end that I'd done so.
